client/kernel-switcher

```diff
@@ -21,21 +22,68 @@ from solo.utils import print_package_info
 import kswitch
 
 
-if __name__ == "__main__":
-
-    app_name = os.path.basename(sys.argv[0])
-    parser = argparse.ArgumentParser(
-        description=_("Sabayon Kernel Switcher"),
-        formatter_class=argparse.RawDescriptionHelpFormatter,
-        prog=app_name)
-
+class SabayonSwitcher(argparse.Namespace):
```

Comment:
What's the advantage to doing it this way?

Frankly I'm seeing it for the first time with argparse to have the main class inherit this. The result here is that `called_method` is melded into the main class instead of being property of the parser.

I see it as a problem because the class now does not only implement the functionality but also does something related to how it's executed (which function - something that comes from the outside).

Hm?

Reply:
Well, SabayonSwitcher class target is the executor of kernel-switcher and permit to create in an easy way script for automatic tests. I prefer use a class instead of pure python script because could be more integrable and more easy to read.

1) Extend of argparse.Namespace is not mandatory. Use of SabayonSwitcher class as namespace could be done without it but I added this to avoid warning from pycharm on use this class (See [this](https://docs.python.org/3/library/argparse.html#the-namespace-object))

2) Use directly SabayonSwitcher as namespace avoid to set local attribute with result of arguments parsing. It's only called_method set automatically but also ask, pretend, kernel, etc.
Why you want do a thing that it's already done automatically from argparse library? I think that this is cool.

3) I don't understand your comment about it's a problem use argparse to automatically set execution parameters. SabayonSwitcher is the binary for kernel switching and logic depends on value of the input parameters.

4) With this simple code:

```python
if __name__ == "__main__":

    app = SabayonSwitcher(
        app_name=os.path.basename(sys.argv[0])
    )
    ret = app.main(sys.argv)
    raise SystemExit(ret)
```

instead of have all execution code under some section you can create test suite with something like this:

```python
args = [ 'kernel-switcher' , 'list' ]
app = SabayonSwitcher()
ret = app.main(args)
```

Reply:
OK, given that it's a class, indeed it's easier to have the parser just inside it, and setting method bound to the instantiated class, self.something.

I was about to say that still the options should be separated, like `self.args.ask`, `self.args.pretend`, and argparse can do it perfectly (which would also fulfill everything you wrote). But then, I noticed that the only attributes are the ones that come from the command line (except the one now prefixed with underscore), so there is no place for confusion. I hope it stays like this.

This is not my style, but seems OK. The problematic thing is that if you were to unit test it, you would be virtually forced to use shell-like arguments (--pretend, --ask) and go though argparse, due to lack of separation, which is kind of silly. Or am I missing something? Do you think it's a problem?

Reply:
PS

> Why you want do a thing that it's already done automatically from argparse library? I think that this is cool.

There seems to be some misunderstanding probably (in the other solution argparse would just create a Namespace object which, even as is, would be passed to functions or set as an attribute).

Reply:
Normally, I use not direct configuration from argparse where there is a configuration file because you need handle priority of parameters and override and yes in this case probably it's better solution where arguments are under self.args.* but in this case, with very few arguments and no configuration file could be fine set directly attributes with argparse. imho.

About unit test, yes in this case you need force a list with shell-like arguments but we can bypass this. Often I use an optional parameter to main function:

```python
def main(parse_cmd_line=True):
    if parse_cmd_line:
          # Do argparse stuff
```

Solution 1:

So in this case test unit became something like this:
```python
s = SabayonSwitcher()
s.ask = False
s.called_method = s._list
s.main(parse_cmd_line=False)
```

Solution 2:
Initialize class and use directly right method and skip main:
```python
s = SabayonSwitcher()
s.ask = False
s.kernel = 'sys-kernel/linux-sabayon-4.16.18'
s._switch()
```


wdyt ?

Reply:
Yeah it's workable, especially the second approach seems OK, even though personally I'd prefer more "functional" approach rather than setting parameters after creating the object.
We can go back and think about this if/when UTs are being made.